File: analysis/beamforming/steeringVectors.py
# -*- coding: utf-8 -*-
"""
Created on Thu Feb 14 11:53:42 2019

@author: bfj
"""
import numpy as np

speedOfLight = 299792458.0

def steeringVectors( coords, angles, freqsInHz, s21data):
    #return hThetaF

    
    #% coords assumed to have centroid 0
    numFreqs = len(freqsInHz)
    numPoints = coords.shape[0]
    numAngles = angles.shape[0]
    
    
    lambdaVec = speedOfLight / (freqsInHz)
    kVec = 2 * np.pi / lambdaVec
    
    #%steeringVectors = zeros(numAngles, numPoints);
    
    dirs = np.zeros((numAngles,3))
    #dirs[:,0] = np.sin(angles[:,0])*np.sin(angles[:,1])
    #dirs[:,1] = np.sin(angles[:,0])*np.cos(angles[:,1])
    #dirs[:,2] = np.cos(angles[:,0])
    dirs[:,0] = np.sin(angles[:,1])
    dirs[:,1] = np.sin(angles[:,0])
    dirs[:,2] = np.sqrt(1.0 - dirs[:,0]**2 + dirs[:,1]**2)
    hThetaF = np.zeros((numAngles, numFreqs), dtype=np.complex128)
    
    #% Could be much faster...
    for iF in range(numFreqs):
        print(iF)
        kDir = kVec[iF]*dirs
        steeringVectorsData = np.exp(-1j * np.dot(kDir,coords.transpose()))
        
        ################################################################
        # Efficient implementation - needs to be verified
        ################################################################
        #num = np.dot(steeringVectorsData, s21data[:,iF])
        ## Tensor notation - calculate the diagonal 
        #denom = np.einsum('ij,ji->i', steeringVectorsData, steeringVectorsData.transpose().conj())
        #hThetaF[:, iF] = num/denom
        ################################################################
        
        ################################################################
        
        ################################################################
        for iA in range(numAngles):
            num = np.dot(np.reshape(s21data[:,iF], -1).transpose() ,np.reshape(steeringVectorsData[iA,:], -1))
            denom = np.dot(np.conj(np.reshape(steeringVectorsData[iA,:], -1)).transpose(),np.reshape(steeringVectorsData[iA,:], -1))
            hThetaF[iA, iF] = num/np.sqrt(denom)
        #end
        
        #% The next line works well...
        #%hThetaF(:,iF) = exp(1j*kVec(iF)*dirs*coords')*conj(s21data(:,iF)); %reshape(s21data(:,iF), [], 1)'*conj(reshape(steeringVectors(iA,:), [], 1));
        #%denom = sum(abs(steeringVectors).^2,2); %    sum(conj(reshape(steeringVectors(iA,:), [], 1)).*reshape(steeringVectors(iA,:), [], 1),1);
        #%hThetaF(:, iF) = num ; num./sqrt(denom);
    #end
    
    #%hThetaF = hThetaF/numPoints;
        
       
    return hThetaF
    

File: samurai/analysis/beamforming/matlab/reverb/steeringVectors.m
function [ hThetaF ] = steeringVectors( coords, angles, freqsInHz, s21data)

% coords assumed to have centroid 0
numFreqs = length(freqsInHz);
numPoints = size(coords,1);
numAngles = size(angles,1);

speedOfLight = 299792458.0;
lambdaVec = speedOfLight ./ (freqsInHz);
kVec = 2 * pi ./ lambdaVec;

%steeringVectors = zeros(numAngles, numPoints);

dirs = zeros(numAngles,3);
dirs(:,1) = cos(angles(:,1)).*cos(angles(:,2));
dirs(:,2) = cos(angles(:,1)).*sin(angles(:,2));
dirs(:,3) = sin(angles(:,1));

hThetaF = complex(zeros(numAngles, numFreqs));

% Could be much faster...
for iF = 1:numFreqs
    kDir = kVec(iF)*dirs;
    steeringVectors(:, :) = exp(-1j * kDir*coords');
    
    
    for iA = 1:numAngles
        num = reshape(s21data(:,iF), 1, [])*reshape(steeringVectors(iA,:), [], 1);
        denom = sum(conj(reshape(steeringVectors(iA,:), [], 1)).*reshape(steeringVectors(iA,:), [], 1),1);
        hThetaF(iA, iF) = num./sqrt(denom);
    end
    
    % The next line works well...
    %hThetaF(:,iF) = exp(1j*kVec(iF)*dirs*coords')*conj(s21data(:,iF)); %reshape(s21data(:,iF), [], 1)'*conj(reshape(steeringVectors(iA,:), [], 1));
    %denom = sum(abs(steeringVectors).^2,2); %    sum(conj(reshape(steeringVectors(iA,:), [], 1)).*reshape(steeringVectors(iA,:), [], 1),1);
    %hThetaF(:, iF) = num ; num./sqrt(denom);
end

%hThetaF = hThetaF/numPoints;
    
    %for iA = 1:numAngles
        %kDir = kVec(iF);
        %steeringVectors(iF, iA, :) = exp(-1j * coords*kDir');
        
        %for iP = 1:numPoints
        %    %coords(iP,:)
        %    %dir = coords(iP,:);
        %    steeringVectors(iF, iA, iP) = exp(1j * sum(coords(iP,:).*kDir));
        %end
    %end


return
%array_z = repmat( RxZ, Nelem, Nelem );

kx = k * xVals;
ky = k * yVals; 
%kz = k * array_z;

numUvPts = 2/duv + 1;
u1 = linspace(-1,1,numUvPts);
v1 = linspace(-1,1,numUvPts);

[uGrid, vGrid] = meshgrid(u1, v1);
steeringVectors = zeros(numUvPts, numUvPts, numPoints);
mask = ones(numUvPts, numUvPts);

for iU = 1 : numUvPts
    for iV = 1 : numUvPts
        if (uGrid(iV, iU)^2 + vGrid(iV, iU)^2) > 1
            steeringVectors(iV, iU, :) = NaN;
            mask(iV, iU) = 0;
            continue
        end
            
        k_az3 = kx * uGrid( iV, iU );
        k_el3 = ky * vGrid( iV, iU );
        steeringVectors(iV, iU, :) = exp( -1j * ( k_az3( : ) + k_el3( : ) ) );
        
     end
end



